bigqueue_flush_test.go

```diff
@@ -0,0 +1,274 @@
+
+	msg := []byte("a")
+	for i := 0; i < arenaSize/cInt64Size; i++ { // a lot of writes, but still no flush
+		if err := bq.Enqueue(msg); err != nil {
+			t.Fatalf("enqueue failed :: %v", err)
+		}
+	}
+
+	checkDirtiness(bq, t, true, [3]bool{true, true, false})
+	checkTimesCalled(bq, t, 0, [3]int{0, 0, 0})
+
+	clock.Advance(flushElapsedDuration)
+
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	checkDirtiness(bq, t, false, [3]bool{false, false, false})
+	checkTimesCalled(bq, t, 1, [3]int{1, 1, 0})
+
+	clock.Advance(time.Second / 2)
+
+	for i := 0; i < arenaSize/cInt64Size; i++ {
+		if err := bq.Enqueue(msg); err != nil {
+			t.Fatalf("enqueue failed :: %v", err)
+		}
+	}
+
+	checkDirtiness(bq, t, true, [3]bool{false, true, true})
+	checkTimesCalled(bq, t, 1, [3]int{1, 1, 0})
+
+	clock.Advance(time.Second)
+
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	checkDirtiness(bq, t, false, [3]bool{false, false, false})
+	checkTimesCalled(bq, t, 2, [3]int{1, 2, 1})
+}
+
+func TestResetOfBothFlushIntervalStates(t *testing.T) {
+	testDir := path.Join(os.TempDir(), fmt.Sprintf("testdir_%d", rand.Intn(1000)))
+	createTestDir(t, testDir)
+	defer deleteTestDir(t, testDir)
+
+	flushElapsedDuration := time.Second
+	clock := clockwork.NewFakeClock()
+
+	bq, err := NewMmapQueue(testDir, SetFlushElapsedDuration(flushElapsedDuration),
+		SetFlushIntervalMutateOps(2),
+		setClock(clock))
+
+	if err != nil {
+		t.Fatalf("unable to get BigQueue: %v", err)
+	}
+	defer bq.Close()
+
+	msg := []byte("a")
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	checkMutateOps(bq, t, 1)
+	clock.Advance(flushElapsedDuration) // time interval reached, flush happens, resets both prevFlushTime, mutateOps
+
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	checkMutateOps(bq, t, 0)
+	prevFlushTime := (bq.(*MmapQueue)).prevFlushTime
+	if prevFlushTime != clock.Now() {
+		t.Fatalf("expected prevFlushTime: %v to be reset to clock's now: %v", prevFlushTime, clock.Now())
+	}
+
+	clock.Advance(flushElapsedDuration / 2) // time moves ahead
+
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+	if err := bq.Enqueue(msg); err != nil {
+		t.Fatalf("enqueue failed :: %v", err)
+	}
+
+	// flush after 2 mutateOps, causes mutateOps reset and time reset to Now()
+
+	checkMutateOps(bq, t, 0)
+	prevFlushTime = (bq.(*MmapQueue)).prevFlushTime
+	if prevFlushTime != clock.Now() {
+		t.Fatalf("expected prevFlushTime: %v to be reset to clock's now: %v", prevFlushTime, clock.Now())
+	}
+
+}
+
+type flushCountingFile struct {
+	mmap.File
+	timesCalled int
+}
+
+func (m *flushCountingFile) Flush(flags int) error {
+	m.timesCalled++
+	return m.File.Flush(flags)
+}
+
+func setupFlushCountFile(q Queue, t *testing.T) {
+	bq := q.(*MmapQueue)
+	for i := 0; i < bq.conf.maxInMemArenas; i++ {
```

Comment:
shouldn't this iterate over all the arenas, instead of just as as many as we are allowed to keep in memory?

Reply:
arenas not in memory would be nil in the arenalist and hence they can't be mocked.
so this test sets up (loads in memory) the first N number of arenas where N = maxInMemArenas.

Reply:
Oh, and then you make sure that you only write to these arenas. I see.